docs/zenith/developer/go/457202-test-build-publish.md

```diff
@@ -0,0 +1,303 @@
+    return dag.Ttlsh().Publish(ctx, m.Package(nodeVersion))
+  }
+  ```
+
+    - The `Package()` function calls the `dag` client's `Container().From()` method to initialize a new container from a base image - here, the `nginx:1.23-alpine` image. The `From()` method returns a new `Container` object with the result. It uses the `Container.WithDirectory()` method to write the `Directory` returned by the `Build()` function to the `/usr/share/nginx/html` path in the container and return a revised `Container`.
+    - The `Publish()` function accepts a `Container` as input. It calls the `ttlsh` module's `Publish()` method to publish the container image to the [ttl.sh](https://ttl.sh) registry and return the image identifier.
+
+Test the code by running the command below:
+
+```shell
+dagger call publish
+```
+
+Here's an example of the output you will see:
+
+```shell
+✔ dagger call publish [1m12.8s]
+┃ ttl.sh/pensive_murdock:10m@sha256:2c483d2e6aec5f950221363aaf1cdde5aceab906dd85d2b63e97acda48881b5a
+```
+
+## Step 6: Test and debug the module
+
+Apart from the usual Go debugging techniques, Dagger provides two commands that come in very handy when debugging modules:
+
+### dagger shell
+
+The `dagger shell` command can be used to open an interactive shell session with any `Container` type returned by a function. This is very useful for debugging and experimenting since it allows you to interact with containers directly.
+
+By default, `dagger shell` will execute the container's entrypoint, but you can override this with the `--entrypoint` flag.
+
+Try it with the command below:
+
+```shell
+dagger shell package --entrypoint /bin/sh
+```
+
+This command drops you into an interactive shell, allowing you to directly execute commands in the container returned by the `package` function, as in the example below:
+
+```shell
+/ # cd /usr/share/nginx/html
+/usr/share/nginx/html # ls
+50x.html             favicon.ico          logo192.png          manifest.json        static
+asset-manifest.json  index.html           logo512.png          robots.txt
+/usr/share/nginx/html #
+```
+
+### dagger up
+
+The `dagger up` command allows `Service` and `Container` types returned by a function to be executed and have any exposed ports forwarded to the host machine. This has many potential use cases, such as manually testing web servers or databases directly from the host browser or host system.
+
+In order for this to work, the service/container returned by the function must have the `Container.withExposedPort` field defining one or more exposed ports. So, add a new `PackageService()` function as shown below:
+
+```go
+func (m *Mymod) PackageService(nodeVersion Optional[string]) *Service {
+	return m.Package().
+		WithExposedPort(8080).
+		AsService()
+}
+```
+
+Then, use the `dagger up` command to build the application and serve it with NGINX:
+
+```shell
+dagger up --native package-service
```

Comment:
This fails for me with a service health check error. Any idea why?

Reply:
It's due to switching the base container image to nginx alpine rather than the chainguard one; the chainguard one is preconfigured to have the default port be 8080.

Is there any reason we switched from chainguard (here https://github.com/dagger/hello-dagger/blob/44e6fd7aaa56997c8a26cac6b950d18c18938f20/main.go#L13) to alpine?

Reply:
Thanks for checking this! I switched to nginx-alpine to try and stay as close as possible to the current quickstart, and also to use a Docker Hub image (which I think is more common?) I've modified the example to expose port 80, which seems to be the default port for the nginx-alpine image.

No real objection to using chainguard though, lmk if you feel that's preferable.

Reply:
I prefer the chainguard image because it's lighter-weight and thus faster to pull.

Reply:
That's interesting. I always thought that the alpine images were the lightest and quickest to pull.